### PR TITLE
adds logic in relationship field validate to add to projection sub re…

### PR DESCRIPTION
**Pull request by ValJed**
[PRO-6535](https://linear.app/apostrophecms/issue/PRO-6535/projection-not-working-with-nested-relationships)

## Summary

When validating relationship fields, if there `withRelationships` and `builders.project` are set, add needed `idsStorage` fields to project in order to be able to retrieve sub relationship.

## New behavior

If `withRelationships` is set to `\true` we take all first level relationships,
```javascript
          _pages: {
            label: 'Pages',
            type: 'relationship',
            withType: 'default-page',
            withRelationships: true,
            builders: {
              project: {
                title: 1,
                _url: 1
              }
            }
          }
```
In this case if the `default-page` has relationships to foodArticles and sportArticles, it will incluce both.
But if foodArticles has a relationship to topics, it won't take it.
this behavior already worked but was not documented, I just made sure it still works with missing projections.

If `withRelationships` is an array with one or multiple levels of relationships using the dot notation. for each one having a projection, we add the missing fields to build these relationships.

```javascript
          _pages: {
            label: 'Rel page',
            type: 'relationship',
            withType: 'default-page',
            withRelationships: [ '_articles._topics' ],
            builders: {
              project: {
                title: 1
              }
            }
          }
```

In this case we will get pages, their articles and the topics of the articles. Even if the projections on each level do not include the needed relationship properties.

⚠️ It might change some existing behavior since if relationship is not explicitly false we take the first level of relationship even if the projection does not include the right field. I would argue that it's not a problematic change since it's the normal behavior when no projection.

## What are the specific steps to test this change?

See Ticket description:
* Create widget with relationship to a page
* create piece
* add relationship from page to piece
In widget, if a project is added to page (even without idsStorage for the sub relationship) it should work.

If you don't want to spend time configuring a project to test it, I created a branch for you:
https://github.com/ValJed/a3-wp/tree/pro-6535-projections-sub-relationships

It reproduces the behavior described in the ticket.

## What kind of change does this PR introduce?

- [X] Bug fix
- [ ] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [X] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [ ] The changelog is updated
- [ ] Related documentation has been updated
- [ ] Related tests have been updated

### CODE REVIEW

File: modules/@apostrophecms/schema/lib/addFieldTypes.js [hunk 1]
```diff
@@ -1061,9 +1061,21 @@ module.exports = (self) => {
...
+      return self.relationshipDriver(
+        req,
+        joinr.byArray,
+        false,
+        objects,
+        field.idsStorage,
+        field.fieldsStorage,
+        field.name,
+        options
+      );
```
Comment (ValJed) [inline modules/@apostrophecms/schema/lib/addFieldTypes.js:1078]: just lint

File: modules/@apostrophecms/schema/index.js [hunk 2]
```diff
@@ -934,7 +934,10 @@ module.exports = {
               _.each(_objects, function (object) {
                 if (object[relationship.name]) {
                   const locale = `${req.locale}:${req.mode}`;
-                  object[relationship.name] = self.apos.util.orderById(object[relationship.idsStorage].map(id => `${id}:${locale}`), object[relationship.name]);
+                  object[relationship.name] = self.apos.util.orderById(
+                    object[relationship.idsStorage].map(id => `${id}:${locale}`),
+                    object[relationship.name]
+                  );
```
Comment (ValJed) [inline modules/@apostrophecms/schema/index.js:940]: LInt only

File: modules/@apostrophecms/doc-type/index.js [hunk 3]
```diff
@@ -2219,15 +2242,21 @@ module.exports = {
...
               const manager = self.apos.doc.getManager(type);
               // Careful, there will be no manager if type was not part of the projection
               if (manager && manager.schema) {
-                await self.apos.schema.relate(query.req, manager.schema, resultsByType[type], query.get('relationships'));
+                await self.apos.schema.relate(
+                  query.req,
+                  manager.schema,
+                  resultsByType[type],
+                  relationships
+                );
```
Comment (ValJed) [inline modules/@apostrophecms/doc-type/index.js:2259]: Lint

File: modules/@apostrophecms/doc-type/index.js [hunk 4]
```diff
@@ -2219,15 +2242,21 @@ module.exports = {
         relationships: {
           def: true,
           async after(results) {
-            if (!query.get('relationships')) {
+            const relationships = query.get('relationships');
+            if (!relationships) {
```
Comment (ValJed) [inline modules/@apostrophecms/doc-type/index.js:2246]: Lint

File: modules/@apostrophecms/doc-type/index.js [hunk 5]
```diff
@@ -1757,6 +1758,32 @@ module.exports = {
...
+              !Object.values(projection).some((val) => !val)
+            ) {
+              // Didn't find any other way, should work
+              const type = query.get('type');
+              const manager = self.apos.doc.getManager(type);
+              const directRelations = Array.isArray(relationships)
+                ? relationships.map(rel => rel.split('.')[0])
+                : [];
+
+              if (manager) {
```
Comment (boutell) [inline modules/@apostrophecms/doc-type/index.js:1774]: 👍

File: modules/@apostrophecms/doc-type/index.js [hunk 6]
```diff
@@ -1757,6 +1758,32 @@ module.exports = {
...
+                for (const field of manager.schema) {
+                  if (field.type !== 'relationship') {
+                    continue;
+                  }
+
+                  if (relationships === true || directRelations.includes(field.name)) {
+                    add.push(field.idsStorage);
```
Comment (boutell) [inline modules/@apostrophecms/doc-type/index.js:1781]: What about `fieldsStorage`? Make sure to add it only if it is truthy.

**Review by boutell — CHANGES_REQUESTED** [review]
Very nice, want to make sure you cover fieldsStorage where appropriate

File: modules/@apostrophecms/doc-type/index.js (hunk 6 above)
Comment (ValJed) [reply modules/@apostrophecms/doc-type/index.js:1781]: Thanks! Added!

**Review by boutell — APPROVED** [review]

**Pull request merged**